src/openvpn/socket.c

```diff
@@ -1082,6 +1088,39 @@ create_socket_udp(struct addrinfo *addrinfo, const unsigned int flags)
     return sd;
 }
 
+#if defined(TARGET_LINUX) && defined(ENABLE_MPTCP)
+socket_descriptor_t
+create_socket_mptcp(struct addrinfo *addrinfo)
+{
+    socket_descriptor_t sd;
+
+    ASSERT(addrinfo);
+    ASSERT(addrinfo->ai_socktype == SOCK_STREAM);
+    addrinfo->ai_protocol = IPPROTO_MPTCP;
+    if ((sd = socket(addrinfo->ai_family, addrinfo->ai_socktype, addrinfo->ai_protocol)) < 0)
```

Comment:
> it might be interesting (in a second commit?) to implement a fallback: create a socket with `IPPROTO_MPTCP` and if it fails with a certain `errno` (`EPROTONOSUPPORT`), use a `goto` to recreate the socket without MPTCP (`proto = addrinfo->ai_protocol`), e.g.
> 
> ```
>     int proto;
> 
>     (...)
> 
>     proto = options->enable_mptcp ? IPPROTO_MPTCP, addrinfo->ai_protocol;
> 
> socket:
>     if ((sd = socket(addrinfo->ai_family, addrinfo->ai_socktype, proto)) < 0)
>     {
>         /* Fallback to TCP */
>         if (proto == IPPROTO_MPTCP && errno == EPROTONOSUPPORT)
>         {
>             msg(M_WARN, "TCP: MPTCP is not supported by the kernel, fallback to TCP");
> 
>             proto = addrinfo->ai_protocol;
>             goto socket;
>         }
> ```



> (I didn't try the code and I guess you need to use something else than `options->enable_mptcp`)



> @obonaventure maybe we should not check the `errno`: if `net.mptcp.enabled` is disabled (or other restrictions via SELinux and others), the errno will be different but still a fallback is possible.
> 
> We could check for other errno's but probably easier to just print a warning and retry in TCP.
> 
> ```c
>     int proto;
> 
>     (...)
> 
>     proto = options->enable_mptcp ? IPPROTO_MPTCP, addrinfo->ai_protocol;
> 
> socket:
>     if ((sd = socket(addrinfo->ai_family, addrinfo->ai_socktype, proto)) < 0)
>     {
>         /* Fallback to TCP */
>         if (proto == IPPROTO_MPTCP)
>         {
>             msg(M_WARN, "TCP: MPTCP is not supported by the kernel, fallback to TCP");
> 
>             proto = addrinfo->ai_protocol;
>             goto socket;
>         }
> ```